### DISCUSSION TITLE
v0.1.4

### DISCUSSION THREAD

**Discussion by YaLTeR** [YaLTeR/niri#272] [Announcements]
Before we begin: [downgrade `xz` to v5.4](https://www.openwall.com/lists/oss-security/2024/03/29/4). Done? Good, let's get to the release.

Niri is a scrollable-tiling Wayland compositor. Windows are arranged in columns on an infinite strip going to the right. Opening a new window never causes existing windows to resize.

Here are the improvements from the last release.

## Block out windows from screencasts

You can now instruct niri to block out certain windows from screencasts with a [window rule](https://github.com/YaLTeR/niri/wiki/Configuration:-Window-Rules#block-out-from). This can be useful for apps like password managers or messengers that you don't want to accidentally show during meetings or streams.

![Screenshot showing a visible Secrets window and an OBS window screencasting the screen where there's a black rectangle instead of the Secrets window.](https://github.com/YaLTeR/niri/assets/1794388/8a703432-e223-4e3e-aa60-70e45e5df573)

The compositor (niri in this case) is the perfect place for this functionality since it is solely responsible for the video frames sent to displays and screencast clients like OBS. Plus, the compositor has the entire window tree, and can selectively block out individual windows while retaining correct layered compositing.

This blocking out also seamlessly works with the built-in screenshot UI. You can capture a screenshot with interactive area selection while seeing all windows normally, and on a screencast, this entire process, including the interactive selection UI, will have the windows correctly blocked out.

https://github.com/YaLTeR/niri/assets/1794388/3c1f7b0b-1571-439b-8d74-fa425196049e

Unfortunately, this kind of tight integration is not possible with third-party screenshot annotation/preview tools. To avoid accidentally showing windows even when using third-party screenshot tools, niri provides a more aggressive mode that blocks out the window from *all* screen capture tools, not just xdg-desktop-portal screencasts.

## Dynamic window rules

Window rules, introduced in the last release, let you adjust behavior for individual windows. For this release, I did the necessary refactors to support *dynamic* window rules that apply continuously to open windows.

The main example of course is the rule to block out windows from screencasts (described above), but you can also [override whether borders draw with a solid background](https://github.com/YaLTeR/niri/wiki/Configuration:-Window-Rules#draw-border-with-background), change the [window size limits](https://github.com/YaLTeR/niri/wiki/Configuration:-Window-Rules#size-overrides) and adjust [window opacity](https://github.com/YaLTeR/niri/wiki/Configuration:-Window-Rules#opacity).

As an example, you can replicate the "inactive windows become semitransparent" effect:

```
window-rule {
    match is-active=false
    opacity 0.9
}
```

![Screenshot showing three apps, the middle one focused, and the ones on the side are semitransparent.](https://github.com/YaLTeR/niri/assets/1794388/5181bd0f-2e97-4cf9-93b5-c0d22e1171e1)

Check out the [window rules wiki page](https://github.com/YaLTeR/niri/wiki/Configuration:-Window-Rules) for a complete list of properties that you can set.

By the way, I also added a little `niri msg focused-window` IPC command that shows you the title and app ID of the focused window, to aid in writing rule matchers.

## Warp mouse to focus & focus follows mouse

Two popular features among tiling WMs, now in niri.

[Warp mouse to focus](https://github.com/YaLTeR/niri/wiki/Configuration:-Input#warp-mouse-to-focus), implemented by @FluxTape  (thanks!), will automatically move the mouse into windows as you focus them. [Focus follows mouse](https://github.com/YaLTeR/niri/wiki/Configuration:-Input#focus-follows-mouse) on the other hand will automatically focus windows under the cursor as you move it around.

These two actually work very well together, give it a try!

https://github.com/YaLTeR/niri/assets/1794388/b243d7d5-3733-4034-8718-f261e3605e6b

## Mouse and touchpad scroll bindings

It is now possible to bind mouse wheel and touchpad scrolls to perform actions or spawn commands. Mouse scrolling activates every wheel "tick", whereas touchpad scrolling emulates scroll "ticks" based on finger distance traveled.

Additionally, you can now set a cooldown for binds to avoid triggering them too often with scrolling.

```
binds {
    Mod+WheelScrollDown cooldown-ms=150 { focus-workspace-down; }
    Mod+WheelScrollUp   cooldown-ms=150 { focus-workspace-up; }

    Mod+TouchpadScrollDown { spawn "wpctl" "set-volume" "@DEFAULT_AUDIO_SINK@" "0.02+"; }
    Mod+TouchpadScrollUp   { spawn "wpctl" "set-volume" "@DEFAULT_AUDIO_SINK@" "0.02-"; }
}
```

Check the [wiki page](https://github.com/YaLTeR/niri/wiki/Configuration:-Key-Bindings#scroll-bindings) for more details.

## `wlr-gamma-control`

Niri now implements the `wlr-gamma-control` protocol that is required for "Night Light" tools like [wlsunset](https://sr.ht/~kennylevinsen/wlsunset/) and [gammastep](https://gitlab.com/chinstrap/gammastep). Thanks @phuhl for working on this!

> [!NOTE]
> As it turns out, setting the gamma is pretty slow, at least on my system. I suggest using [wlsunset](https://sr.ht/~kennylevinsen/wlsunset/) because it only sets the gamma when it needs to change, rather than, say, every few seconds unconditionally.

## xdg-desktop-portal-gnome 46.0

I implemented more of the Mutter D-Bus interface to support xdg-desktop-portal-gnome 46.0 with its new screencast monitor selector with visual positions:

![Screenshot of the new monitor selector in xdg-desktop-portal-gnome 46.0.](https://github.com/YaLTeR/niri/assets/1794388/79e94d9d-7825-496f-af33-1f2a9e3b51de)

## Documentation

As you've noticed, I've been linking the wiki pages a lot. The reason for this is that over the past week I went through the entire config and wrote detailed documentation and examples for every single option.

Check out all this documentation here on the wiki: https://github.com/YaLTeR/niri/wiki/Configuration:-Overview

A major benefit of this is that it allowed me to declutter the [default config](https://github.com/YaLTeR/niri/blob/v0.1.4/resources/default-config.kdl) by removing some of the less important things and instead linking the wiki. This will improve the experience for people trying out niri as they no longer need to sift through the entire window rule and animation examples.

If you find a mistake, feel free to open a pull request against the `wiki/` folder of the repository.

Also, shoutouts to great reference-style docs @sodiboo had been writing for the nix flake: https://github.com/sodiboo/niri-flake/blob/main/docs.md#programsnirisettings

## Other improvements in this release

- When opening and closing a window without switching focus in between (think various dialogs and temporary windows), niri focuses the previous window since that's where you came from. Now, when this happens, niri will also restore the view position, which makes the behavior more natural and less annoying.
- Fixed wp-viewporter bugs in Smithay (thanks @cmeissl). Particularly, this prevented Chromium and Electron applications from accepting mouse input after resizing.
- Added more information to `niri msg outputs` (logical output position, size, scale, transform, and current and preferred mode flags).
- Added a [`click-method`](https://github.com/YaLTeR/niri/wiki/Configuration:-Input#pointing-devices) input setting for touchpads (thanks @uetcis).
- Added a [`workspace-auto-back-and-forth`](https://github.com/YaLTeR/niri/wiki/Configuration:-Input#workspace-auto-back-and-forth) setting that causes switching to the same workspace by index twice to switch back to the previous workspace (thanks @FluxTape).
- Added support for `ISO_Level3_Shift` / `Mod5` modifier (thanks @Trundle).
- Added a once-per-second fallback timer that sends frame callbacks to off-screen windows, which fixes issues with vsynced games in [gamescope](https://github.com/ValveSoftware/gamescope).
- The last folder in the screenshot path is now automatically created if it doesn't exist.
- Corrected pointer location reported to lock screen surfaces.
- Fixed niri crashing when a screencast is attempted after failing to initialize PipeWire.
    - If on your system PipeWire is not started automatically, you need to make sure that it is started *before* niri for niri to have screencast support.
- Fixed a crash that could happen when stopping the same screencast session twice with the right timing.
- Relaxed checks for DRM render nodes, which in theory allows niri to run on more devices with split DRM display/render nodes (various ARM boards like Raspberry Pi).
- Made the EGL wl-display extension optional on the TTY which makes niri work on some NVIDIA GPUs where it didn't before.
- Niri now tries to reduce the max bits-per-channel output property to 8 which may result in more monitor configurations working.
- Fixed building on musl.
- Fixed mouse scrolling inside nested niri window sometimes being too slow.
- Fixed hardcoding `us` as the keyboard layout when it is missing from the config (turns out libxkbcommon handles that one for us).

<hr /><em>This discussion was created from the release <a href='https://github.com/YaLTeR/niri/releases/tag/v0.1.4'>v0.1.4</a>.</em>

**Comment by jgregoire** [comment]
Nice update!

I just used `niri msg focused-window` to write a custom waybar plugin that displays the focused window title:
https://github.com/jgregoire/dotfiles/commit/bbc76d47f94d99d7df9e18797cfa1aeb3e22bf62

**Reply by YaLTeR** [reply]
Instead of using sed, it's probably a better idea to use the JSON output with jq: `niri msg -j focused-window | jq .title` (plus some filtering for null / quotes).

**Reply by sodiboo** [reply]
filtering for quotes? try `jq --raw-output` (`-r`): `niri msg -j focused-window | jq -r ".title | select(. != null)"`

**Reply by jgregoire** [reply]
I guess I'm just old school :)

**Reply by YaLTeR** [reply]
It's not explicitly written anywhere yet, but: I am trying to preserve some backwards-compatibility on the JSON output (i.e. fields will be added but not removed or renamed), but I may adjust the human-readable output if necessary. So that's another reason to prefer the JSON output for scripts.

**Reply by jgregoire** [reply]
Makes sense!

**Comment by cybergaz** [comment]
bro , just add a feature to support environment variables inside the config file ,
it's so annoying
i can't make it dynamic
i have to hard code everything and change frequently

and blur support
and adjustable rounded corner

these are like "must have" features, I'm waiting for months

**Reply by jgregoire** [reply]
Did you forget how much you paid for this software?

**Reply by YaLTeR** [reply]
Not interested in conversing in this tone, but for anybody else who comes across: you can set environment variables for child process in the config since v0.1.3: https://github.com/YaLTeR/niri/wiki/Configuration:-Miscellaneous#environment

**Reply by cybergaz** [reply]
sorry for making everyone feel so disturbed by my comment,
i did not intend anything like that.

@YaLTeR 
> you can set environment variables for child process in the config since v0.1.3:

i was not talking about setting up ENV vars but use of env vars inside config.
for example :
for a keybind say : Mod+S { spawn "sh" "-c" "$HOME/scripts/script.sh"; }

it is because , i am facing problem setting up my PATH variable,
i have to change PATH variable everytime,
because due current configuration option for ENV Vars it `replace` PATH var entirely

if i could do something like
PATH "$HOME/.cargo/bin:$HOME/.local/bin:$HOME/.bun/bin:$HOME/.config/composer/vendor/bin:$PATH"

it would not overwrite or replace PATH var , but append to the existing one.

if there's any solution available already , please edify me.

**Reply by sodiboo** [reply]
firstly, for your keybind:

instead of doing `sh -c $HOME/scripts/script.sh`, you should just spawn `~/scripts/script.sh`. niri will expand a tilde in the zeroeth ("program name") argument of `spawn` (i.e. `spawn "~/scripts/script.sh"` will expand to `$HOME` but `spawn "sh" "-c" "~/scripts/script.sh"` will pass a tilde to the `sh` invocation). additionally, running scripts like this is natively supported on Linux if you set a shebang and make sure the script is marked as executable (`chmod +x ~/scripts/script.sh`)

secondly, for your path:

are you sure you want to append to `PATH` in niri's config? \
are you sure you don't want to instead do that from a .bashrc or equivalent? \
*using the niri config for this is **not** the correct way to do it.*

---

If you really must do it in niri, I suggest you write your config with some templating language and create a wrapper around niri that will automatically fill in the blanks and write it to a tmp file, then launch niri with a config from the tmp file

something like this:

```sh
TMP_CONFIG="$(mktemp --suffix .config.kdl)"
substituteAll ~/.config/niri/config.kdl "$TMP_CONFIG"
niri -c "$TMP_CONFIG"
rm "$TMP_CONFIG"
```

the `substituteAll` function here is from nixpkgs ([definition](https://github.com/NixOS/nixpkgs/blob/ebde306504e278e4ab153474786091f9ffdfa1e4/pkgs/stdenv/generic/setup.sh#L955-L963)) \
ultimately, it will replace a pattern of `@VARNAME@` will the value of `VARNAME` ([implementation](https://github.com/NixOS/nixpkgs/blob/ebde306504e278e4ab153474786091f9ffdfa1e4/pkgs/stdenv/generic/setup.sh#L869-L870))

you can use any substitution function. with this one, you could set variables in your niri config under `environment {` to something like `PATH "@PATH@:@HOME@/.cargo/bin:@HOME@/.local.bin"`

(note that `substituteAll` doesn't actually support uppercase characters at the start, so you'd rather have to do like `substitute` with flags at the end `--subst-var HOME --subst-var PATH` or modify the `awk` filtering to allow uppercase first characters)

bonus: hot reloading courtesy of [`entr`](http://eradman.com/entrproject/)

```sh
echo "̃~/.config/niri/config.kdl" | entr substituteAll ~/.config/niri/config.kdl "$TMP_CONFIG" &
```

it probably doesn't close that entr process properly when niri exits. oh well. you could spawn that whole thing under `spawn-at-startup` too if you want. terminating the forked command is left as an exercise for the viewer

**Comment by horhenaniB55** [comment]
Opacity window rule for Chromium browser doesn't work. Why?

**Reply by YaLTeR** [reply]
You might be hitting https://github.com/YaLTeR/niri/wiki/FAQ#why-is-the-borderfocus-ring-showing-up-through-semitransparent-windows

Why comment on an ancient release?

**Reply by horhenaniB55** [reply]
window-rule {
   match app-id="chromium"
   opacity 0.6
}

This opacity rule is not working, but works fine with other apps? Am I missing something?

**Reply by YaLTeR** [reply]
What does `niri msg windows` say about chromium? Also send a screenshot?

**Reply by horhenaniB55** [reply]
Window ID 81:
  Title: "New Tab - Chromium"
  App ID: "chromium"
  Is floating: no
  PID: 2043
  Workspace ID: 8
  
  
![image](https://github.com/user-attachments/assets/2a5ff8f8-1d4b-4ab6-b014-2f79acb08ed6)

**Reply by YaLTeR** [reply]
Seems ok. Are you sure you're not overriding it with a subsequent window rule?

**Reply by horhenaniB55** [reply]
I'm sure I am not overriding it, I only have 3 window rules as of the moment.

**Reply by horhenaniB55** [reply]
![image](https://github.com/user-attachments/assets/61c610ef-10ee-4817-a02a-7c60e6c513a4)

![image](https://github.com/user-attachments/assets/0a186f18-3ebc-4955-8aa9-b6130f4205ba)

**Reply by horhenaniB55** [reply]
Ah, the focus ring and border. When I use only border window rule opacity doesn't work with chromium but when I disable border and use focus ring the opacity works.

**Reply by YaLTeR** [reply]
So the FAQ entry that I mentioned?..

**Reply by horhenaniB55** [reply]
Focused
![image](https://github.com/user-attachments/assets/eee736b7-6616-4fd4-8e2f-be4614d03bd3)

Unfocused
![image](https://github.com/user-attachments/assets/5b7485e2-e961-499c-9428-e3aedba49a40)


Well, I still have this problem.

**Reply by YaLTeR** [reply]
Please read the FAQ entry that I linked... Use `draw-border-with-background false` window rule
